Noise in @std/expect's toMatchObject diff (vs. Jest's expect): object keys not present in the expected object

On failed `expect(actual).toMatchObject(expected)` assertions, Deno's differ dumps *all* keys in the `actual` object as `-` lines, making it very hard to spot the actual difference. Jest's differ includes only the keys specified in the `expected` object, and for extra array elements, it marks them with a `+`.

**Steps to Reproduce**

1. Clone https://github.com/dandv/Deno-std-expect-toMatchObject-diff-noise.git
2. Run `deno task test`

**Expected behavior**

Jest's `expect` failure output is scoped to the keys in `expected`:

```diff
error: Error: expect(received).toMatchObject(expected)

- Expected  - 0
+ Received  + 4

  Object {
    "author": Object {
      "name": "Ada Lovelace",
      "stats": Object {
        "posts": 42,
      },
    },
    "featured": true,
    "stats": Object {
      "likes": 812,
    },
    "tags": Array [
      "deno",
+     "typescript",
+     "javascript",
+     "runtime",
+     "tutorial",
    ],
    "title": "Getting Started with Deno",
  }
```

The four extra array elements are marked with `+` so the problem is easy to identify.

---

Deno's `@std/expect` dumps the entire **`actual`** object:

```diff
error: AssertionError: Values are not equal.


    [Diff] Actual / Expected


    {
-     allowComments: true,
      author: {
-       avatar: "https://example.dev/avatars/ada.png",
-       bio: "Pioneering developer and runtime enthusiast. Writes about compilers, runtimes, and elegant APIs.",
-       email: "ada@example.dev",
-       id: "u9876",
        name: "Ada Lovelace",
-       social: {
-         github: "ada-lovelace",
-         linkedin: "ada-lovelace-dev",
-         twitter: "@ada_codes",
-       },
        stats: {
-         followers: 1337,
-         following: 88,
          posts: 42,
        },
      },
-     body: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
-     category: {
-       description: "Server-side development and infrastructure topics.",
-       id: "cat-3",
-       name: "Backend",
-       slug: "backend",
-     },
      featured: true,
-     id: "a1b2c3d4-e5f6-7890-abcd-ef1234567890",
-     publishedAt: "2025-06-01T09:00:00Z",
-     readingTime: 7,
-     relatedArticles: [
-       {
-         id: "a0001",
-         slug: "deno-vs-node",
-         title: "Deno vs Node.js",
-       },
-       {
-         id: "a0002",
-         slug: "typescript-basics",
-         title: "TypeScript Basics",
-       },
-       {
-         id: "a0003",
-         slug: "modern-js-tooling",
-         title: "Modern JS Tooling",
-       },
-     ],
-     seo: {
-       canonicalUrl: "https://example.dev/blog/getting-started-with-deno",
-       metaDescription: "Learn the basics of Deno runtime in this hands-on guide.",
-       metaTitle: "Getting Started with Deno | Dev Blog",
-       openGraph: {
-         image: "https://example.dev/og/deno-guide.png",
-         type: "article",
-       },
-     },
-     slug: "getting-started-with-deno",
      stats: {
-       bookmarks: 309,
-       comments: 134,
        likes: 812,
-       shares: {
-         hackernews: 66,
-         linkedin: 87,
-         reddit: 153,
-         twitter: 204,
-       },
-       views: 29041,
      },
-     subtitle: "A practical guide to the modern JavaScript runtime",
      tags: [
        "deno",
-       "typescript",
-       "javascript",
-       "runtime",
-       "tutorial",
      ],
      title: "Getting Started with Deno",
-     updatedAt: "2025-06-12T16:45:00Z",
    }
```

Every property of the received object is dumped, including fields that were never mentioned in `expected` (e.g. `seo`, `relatedArticles`, `body`, `category`, `slug`, the full `author.social` block). All these appear as `-` removals, because the differ is comparing the full received value against the partial expected value.

The actual mismatch (`tags`) is buried near the bottom.


**Environment**

<!-- please complete the following information -->

- OS:  Ubuntu 24.04
- deno version: 2.7.11
- std version: 1.0.18

--- a/expect/_matchers.ts
+++ b/expect/_matchers.ts
@@ -21,6 +21,7 @@ import { getMockCalls } from "./_mock_util.ts";
 import { inspectArg, inspectArgs } from "./_inspect_args.ts";
 import {
   buildEqualOptions,
+  getObjectSubset,
   iterableEquality,
   subsetEquality,
 } from "./_utils.ts";
@@ -601,7 +602,8 @@ export function toMatchObject(
           : defaultMessage,
       );
     } else {
-      const defaultMessage = buildEqualErrorMessage(received, expected);
+      const subset = getObjectSubset(received, expected, context.customTesters);
+      const defaultMessage = buildEqualErrorMessage(subset, expected);
       throw new AssertionError(
         context.customMessage
           ? `${context.customMessage}: ${defaultMessage}`

--- a/expect/_to_match_object_test.ts
+++ b/expect/_to_match_object_test.ts
@@ -227,4 +227,21 @@ Deno.test("expect().toMatchObject() displays a diff", async (t) => {
       );
     },
   );
+
+  await t.step("omits keys not in expected", () => {
+    const e = assertThrows(
+      () =>
+        expect({ a: 1, b: 2, c: 3, d: { e: 4, f: 5 } }).toMatchObject({
+          a: 1,
+          d: { e: 999 },
+        }),
+      AssertionError,
+    );
+    // The diff should mention the mismatched value
+    assertMatch(e.message, /999/);
+    // The diff should NOT mention keys absent from expected
+    assertNotMatch(e.message, /b:/);
+    assertNotMatch(e.message, /c:/);
+    assertNotMatch(e.message, /f:/);
+  });
 });

--- a/expect/_to_match_object_test.ts
+++ b/expect/_to_match_object_test.ts
@@ -244,4 +244,58 @@ Deno.test("expect().toMatchObject() displays a diff", async (t) => {
     assertNotMatch(e.message, /c:/);
     assertNotMatch(e.message, /f:/);
   });
+
+  await t.step("omits keys not in expected with arrays", () => {
+    const e = assertThrows(
+      () =>
+        expect([{ a: 1, extra: true }, { b: 2 }]).toMatchObject([
+          { a: 999 },
+          { b: 2 },
+        ]),
+      AssertionError,
+    );
+    assertMatch(e.message, /999/);
+    assertNotMatch(e.message, /extra/);
+  });
+
+  await t.step("omits keys not in expected with Date values", () => {
+    const e = assertThrows(
+      () =>
+        expect({
+          d: new Date("2020-01-01"),
+          extra: "noise",
+        }).toMatchObject({ d: new Date("2025-01-01") }),
+      AssertionError,
+    );
+    assertNotMatch(e.message, /noise/);
+  });
+
+  await t.step("omits keys not in expected with equal nested subset", () => {
+    const e = assertThrows(
+      () =>
+        expect({
+          a: { x: 1, y: 2 },
+          b: 999,
+          extra: true,
+        }).toMatchObject({ a: { x: 1 }, b: 42 }),
+      AssertionError,
+    );
+    // b is the mismatch
+    assertMatch(e.message, /999/);
+    assertMatch(e.message, /42/);
+    // extra top-level key should be omitted
+    assertNotMatch(e.message, /extra/);
+    // y should be omitted (not in expected.a)
+    assertNotMatch(e.message, /y:/);
+  });
+
+  await t.step("handles circular references without throwing", () => {
+    const received: Record<string, unknown> = { a: 1 };
+    received.self = received;
+    const e = assertThrows(
+      () => expect(received).toMatchObject({ a: 999, self: {} }),
+      AssertionError,
+    );
+    assertMatch(e.message, /999/);
+  });
 });

--- a/expect/_utils.ts
+++ b/expect/_utils.ts
@@ -281,3 +281,53 @@ export function subsetEquality(
 
   return subsetEqualityWithContext()(object, subset);
 }
+
+// Ported from https://github.com/jestjs/jest/blob/442c7f692e3a92f14a2fb56c1737b26fc663a0ef/packages/expect-utils/src/utils.ts#L82
+export function getObjectSubset(
+  object: unknown,
+  subset: unknown,
+  customTesters: Tester[] = [],
+  seenReferences: WeakMap<object, boolean> = new WeakMap(),
+): unknown {
+  if (Array.isArray(object)) {
+    if (Array.isArray(subset) && subset.length === object.length) {
+      return subset.map((_: unknown, i: number) =>
+        getObjectSubset(object[i], subset[i], customTesters, seenReferences)
+      );
+    }
+  } else if (object instanceof Date) {
+    return object;
+  } else if (isObject(object) && isObject(subset)) {
+    if (
+      equal(object, subset, {
+        customTesters: [...customTesters, iterableEquality, subsetEquality],
+      })
+    ) {
+      return subset;
+    }
+
+    const obj = object as Record<string, unknown>;
+    const sub = subset as Record<string, unknown>;
+    const trimmed: Record<string, unknown> = {};
+    seenReferences.set(object as object, true);
+
+    for (const key of Object.keys(obj)) {
+      if (!Object.prototype.hasOwnProperty.call(sub, key)) continue;
+
+      const val = obj[key];
+      if (typeof val === "object" && val !== null && seenReferences.has(val)) {
+        trimmed[key] = val;
+      } else {
+        trimmed[key] = getObjectSubset(
+          val,
+          sub[key],
+          customTesters,
+          seenReferences,
+        );
+      }
+    }
+
+    if (Object.keys(trimmed).length > 0) return trimmed;
+  }
+  return object;
+}

